aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -1,57 +1,127 @@
 #include "ATen/cuda/CUDAStream.h"
 #include "ATen/cuda/CUDAContext.h"
 #include "ATen/cuda/Exceptions.h"
+#include "ATen/DeviceGuard.h"
 #include "ATen/Error.h"
 
 #include <mutex>
 #include <atomic>
+#include <cstdint>
+#include <deque>
+#include <vector>
 
 // Internal implementation is entirely hidden
+// Note: CUDAStreamInternals doubles for a THCStream
 struct CUDAStreamInternals {
-  bool is_destructible;
-  std::atomic<int> refcount;
-  int64_t device; // Note: cudaGetDevice works with int32_t, not int64_t
-  cudaStream_t stream;
+  CUDAStreamInternals() = default;
+
+  ~CUDAStreamInternals() {
+    if (stream) cudaStreamDestroy(stream);
+  }
+
+  int64_t device = -1; 
+  cudaStream_t stream = nullptr;
 };
 
 namespace at {
 namespace cuda {
 
 namespace detail {
 
-  /*
-  * Stream state
-  */
-  static constexpr cudaStream_t DEFAULT_STREAM = 0;
-
-  static std::once_flag init_flag;
+  // Global stream state and constants
   static int64_t num_gpus;
-  static CUDAStreamInternals* default_streams;
+  static constexpr int STREAMS_PER_POOL = 32;
+  static constexpr unsigned int DEFAULT_FLAGS = cudaStreamNonBlocking;
+  static int HIGH_PRIORITY = 0;
```

Comment:
btw if you're already touching this, our convention in ATen/PyTorch for global constants is `kCamelCaseConstant`, might want to change this

Reply:
Sounds good.

Reply:
Hmm... can we initialize these constants to something that is more obviously uninitialized? It's also a bit annoying that they look like constants, but they are not valid until `initDeviceStreamState` is called... this should be documented here, at least.

Reply:
Since these constants are not initialized for HIP, I'd quite like it if they were ifdef'ed out, so people don't accidentally use them (in their uninitialized form).